### ISSUE TITLE
Use ImageBitmaps for GLTFLoader (and possibly others)

### ISSUE THREAD

**Issue by toji** [mrdoob/three.js#19511]
**TL;DR:** I'd like to propose that ImageBitmapLoader be used by default (with a fallback to ImageLoader for unsupported browsers) for at least the glTF loader and possibly others that have appropriate restrictions on their input images. 

I've been doing some experiments to try and reduce framerate hitches in https://xrdinosaurs.com, especially in VR. The biggest issues I've seen are that when you load in a new model (all binary glTF files with Draco compression and embedded images) there was always a long and noticable hitch in the framerate immediately before the newly loaded dinosaur displayed for the first time.

Turns out one of the biggest blocking operations on the main thread is the internal image decode that happens in texImage2D. Here's a trace of a similar situation recorded from the `webgl_loader_gltf.html` example:

<img width="666" alt="imageLoaderProfile" src="https://user-images.githubusercontent.com/805273/83335515-2df8db80-a262-11ea-9f36-f4d4b66e6945.png">

(Measurements taken on my desktop PC: Intel Core i5 3.3Ghz with a GTX 1660 GPU)

The really big column to the right is the first frame where the loaded glTF is going to be displayed. Some of the largest time sinks in that rather large operation are the dark green blocks at the bottom, which are the aforementioned image decodes. They take **103.29ms**, **24.81ms**, **30.23ms**, **16.07ms** respectively, each of which is blocking the main thread. The entire task (frame) takes **448.22ms** (28 frames @ 60Hz, 40 frames @ 90Hz VR). This is enough to cause a noticeable stall in any VR experience.

[ImageBitmaps](https://developer.mozilla.org/en-US/docs/Web/API/ImageBitmap), if you're not already familiar with them, were designed to allow image decoding to happen pre-emptively off the main thread to produce images that are "ready to use" by the GPU. `gl.texImage2D()` accepts them, and will frequently take significantly less time to complete because the decode step doesn't need to happen synchronously.

I made a simple switch inside `TextureLoader.load()` to allow `ImageBitmapLoader` to be used instead of `ImageLoader` if supported:

```js
// Current code
//var loader = new ImageLoader( this.manager );

// New code
var loader = null;
if (typeof createImageBitmap !== 'unknown') {
  loader = new ImageBitmapLoader( this.manager );
} else {
  // Because some browser still don't support ImageBitmaps.
  loader = new ImageLoader( this.manager );
}	
```

And then ran the same example again:

<img width="701" alt="imageBitmapLoaderProfile" src="https://user-images.githubusercontent.com/805273/83335815-3eaa5100-a264-11ea-9212-91de211eb64d.png">

You can see now that the rightmost task is much shorter now (**260.13ms** in total, nearly cut in half!) The image decoding blocks are still there, but they're happening in a thread pool off the main thread. The initial frame the mesh is first displayed in is delayed because it's waiting on the decode in the other thread, but because that frame now completes faster the user perception is that it takes about the same amount of time with the main difference being that the scene stays more responsive while the load is happening. The difference is more pronounced on https://xrdinosaurs.com where the stall now registers as a minor blip rather than an uncomfortable and lengthy stall.

Of course, nothing is ever that simple, is it? Turns out this only works for glTF files or similar formats where the texture usage is well defined. In other cases where the textures are being applied to dynamically generated meshes they have an annoying tendency to come in upside down because `ImageBitmap` texture sources ignore the `gl.UNPACK_FLIP_Y_WEBGL` parameter, and the `flipY` option that they do support must be set at load time, while `THREE.Texture` allows it to be set at any time. As a result a straight universal replacement of `ImageLoader` doesn't work. In the context of `GLTFLoader` specifically, however, we know that the textures will never need to be flipped because the format dictates that's the case. So it's safe to use `ImageBitmapLoader` in that context while continuing to use `ImageLoader` everywhere else. I imagine this would be the case for many model formats, really, I just don't have a personal motivation to examine anything other than glTF at the moment.

You can see [the code change I applied to my project here](https://github.com/toji/xr-dinosaurs/commit/ffd1a4118397ac9f9e54e8631ca2d470696d48bb).

**Comment by toji** [comment]
BTW: I'm happy to put up a pull request to make this change, but I wanted to ensure it was something that the maintainers approved of first and that any potential pitfalls or alternate approaches could be discussed.

**Comment by donmccurdy** [comment]
We investigated this a few years ago (https://github.com/mrdoob/three.js/pull/12456). At the time, support was good in Chrome, but (a) significantly limited in Firefox, and (b) quite broken in Safari. I am able to see https://threejs.org/examples/?q=imagebitmap#webgl_loader_imagebitmap in both Safari and Firefox now, so perhaps this has improved. Do you know which browsers your GLTFLoader change would support?

> So it's safe to use ImageBitmapLoader in that context while continuing to use ImageLoader everywhere else. I imagine this would be the case for many model formats...

Searching `examples/js/loaders` for occurrences of `flipY`, it appears to be only GLTFLoader and MMDLoader that can guarantee `flipY=false`.

**Comment by gkjohnson** [comment]
I ran into this in my project, as well. This was my solution:

```js
if ( useImageBitmap ) {

  manager.addHandler( /(^blob:)|(\.png$)|(\.jpg$)|(\.jpeg$)/g, {

    load( url, onComplete ) {

      const loader = new ImageBitmapLoader();
      loader.load( url, res => {

        onComplete( new CanvasTexture( res ) );

      } );

    }

  } );

}

new GLTFLoader( manager ).load( ... );
```

I saw really nice performance improvements, as well. Good to hear that there aren't discrepancies across browsers when using it with GLTF. It would be great imo to have this built into GLTFLoader with a graceful fallback like you have.

**Comment by toji** [comment]
> Do you know which browsers your GLTFLoader change would support?

It's working for me in Chromium based browsers and Firefox. I thought it was working on iOS as well, but having just checked it again I'm now seeing some failures. (May have been lulled into a false sense of security by the browser cache. Wouldn't be the first time.) I'll try to do some additional debugging on that. Regardless, this may still be worth investigating on a whitelist basis since the improvement can be so dramatic.

> Searching examples/js/loaders for occurrences of flipY, it appears to be only GLTFLoader and MMDLoader that can guarantee flipY=false

That's mildly surprising. But also, doesn't MMD primarily use TGAs? If so I imagine that `ImageBitmap`s wouldn't apply because you'd be creating `DataTexture`s out of the JS-decoded image.

**Comment by mrdoob** [comment]
> Turns out this only works for glTF files or similar formats where the texture usage is well defined. In other cases where the textures are being applied to dynamically generated meshes they have an annoying tendency to come in upside down because ImageBitmap texture sources ignore the gl.UNPACK_FLIP_Y_WEBGL parameter, and the flipY option that they do support must be set at load time, while THREE.Texture allows it to be set at any time.

Maybe it's time to move away from `gl.UNPACK_FLIP_Y_WEBGL` and do the flipping in glsl: `1.0 - uv.y`.

**Comment by takahirox** [comment]
Sounds good to me that we use ImageBitmapLoader in case where it works fine without any problems. In addition to the performance improvement, it has advantage that it works in Workers.

**Comment by toji** [comment]
Okay, so my Safari issue was due to a [really dumb mistake](https://github.com/toji/xr-dinosaurs/commit/f7a737092c8f1201dbc0c3f49efb7337880e11e6) on my part. 😓 

Good news is, though, that after fixing that my code appears to be working on every browser I've tested.

**Comment by mrdoob** [comment]
> Okay, so my Safari issue was due to a [really dumb mistake](https://github.com/toji/xr-dinosaurs/commit/f7a737092c8f1201dbc0c3f49efb7337880e11e6) on my part. 😓 

Haha, I saw the code and I thought... `unknown`? Yet another JS thing I've never heard of 😁

**Comment by takahirox** [comment]
> Maybe it's time to move away from gl.UNPACK_FLIP_Y_WEBGL and do the flipping in glsl: 1.0 - uv.y.

Probably sounds a good idea. It would resolve this issue, too #16144
